File: app/src/androidTest/java/com/elyeproj/javadaggeruitest/TestComponentRule.java
package com.elyeproj.javadaggeruitest;

import android.content.Context;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class TestComponentRule implements TestRule {

    private ApplicationTestComponent mTestComponent;
    private Context mContext;

    public TestComponentRule(Context context) {
        mContext = context;
    }

    public Context getContext() {
        return mContext;
    }

    public InjectedData getMockInjectedData() {
        return mTestComponent.dataManager();
    }

    private void setupDaggerTestComponentInApplication() {
        MainApplication application = MainApplication.Companion.get(mContext);
        mTestComponent = DaggerApplicationTestComponent.builder()
                .applicationTestModule(new ApplicationTestModule(application))
                .build();
        application.setComponent(mTestComponent);
    }

    @Override
    public Statement apply(final Statement base, Description description) {
        return new Statement() {
            @Override
            public void evaluate() throws Throwable {
                try {
                    setupDaggerTestComponentInApplication();
                    base.evaluate();
                } finally {
                    mTestComponent = null;
                }
            }
        };
    }
}

File: app/src/androidTest/java/com/elyeproj/kotlinespressomock/TestComponentRule.kt
package com.elyeproj.kotlinespressomock

import android.content.Context

import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement

class TestComponentRule(val context: Context) : TestRule {

    private var mTestComponent: ApplicationTestComponent? = null

    val mockInjectedData: InjectedData
        get() = mTestComponent!!.dataManager()

    private fun setupDaggerTestComponentInApplication() {
        val application = MainApplication[context]
        mTestComponent = DaggerApplicationTestComponent.builder().applicationTestModule(ApplicationTestModule(application)).build()
        application.component = mTestComponent as ApplicationComponent
    }

    override fun apply(base: Statement, description: Description): Statement {
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                try {
                    setupDaggerTestComponentInApplication()
                    base.evaluate()
                } finally {
                    mTestComponent = null
                }
            }
        }
    }
}
